[android] - do not update marker while creating

tobrun opened the pull request:
closes #6208 
closes #5915

This PR removes unnecessary `updateAnnotation` calls. 
This was resulting in having ghost annotations. 
This PR also adds a test activity to prevent regression.

Review @ivovandongen / @zugaldia 
cc @jfirebaugh

jfirebaugh commented:
As a gl-core exercise, how about also adding asserts to `AnnotationManager` that trigger on attempts to update or remove a non-existent annotation? That probably would have caught this issue sooner.

tobrun commented:
@jfirebaugh I added asserts in 722053c.

For updating I'm checking if a key is available:
 -  `assert(symbolAnnotations.count(id));`

For removing I'm failing when a key is not available:
 -  `assert(false)` after an else check

jfirebaugh commented:
Asserts look good. However `addAnnotation` will require some refactoring to avoid triggering them. Instead of having it call `updateAnnotation`, try:

```
    return Annotation::visit(annotation, [&] (const auto& annotation_) {
        return this->add(id, annotation_, maxZoom);
    });
```

And then update `AnnotationManager::update` to make it clear the `it == symbolAnnotations.end()` case is not expected.

tobrun commented:
Adapted the addAnnotation to:

```cpp
AnnotationID AnnotationManager::addAnnotation(const Annotation& annotation, const uint8_t maxZoom) {
    AnnotationID id = nextID++;
    Annotation::visit(annotation, [&] (const auto& annotation_) {
        this->add(id, annotation_, maxZoom);
    });
    return id;
}
```
The return type is AnnotationID vs Update in https://github.com/mapbox/mapbox-gl-native/pull/6314#issuecomment-246879845. Is this a valid construction?

jfirebaugh commented on src/mbgl/annotation/annotation_manager.cpp:
```diff
@@ -72,6 +76,7 @@ void AnnotationManager::add(const AnnotationID& id, const StyleSourcedAnnotation
 }
 
 Update AnnotationManager::update(const AnnotationID& id, const SymbolAnnotation& annotation, const uint8_t maxZoom) {
+    assert(symbolAnnotations.count(id));
```
`AnnotationManager::update` used to be expected to have the ability to add new annotations, but now it doesn't. Let's make this clear by changing it to:

```C++
 Update AnnotationManager::update(const AnnotationID& id, const SymbolAnnotation& annotation, const uint8_t maxZoom) {
     auto it = symbolAnnotations.find(id);
     if (it == symbolAnnotations.end()) {
         assert(false); // Attempt to update a non-existent symbol annotation
         return Update::Nothing;
     }
     ...
```

ivovandongen commented:
Besides CI failing, Android side looks good.

tobrun commented:
@jfirebaugh assertions have been changed as requested in 027f607. 
I also took the time to add some unit tests around this in 0a98ca4.

jfirebaugh commented:
I appreciate the thoroughness! However we don't usually test assertions -- assertions are themselves sort of a runtime test, tend to change frequently, and their exact message is platform specific, which I think is why CI is failing. Let's go ahead and leave those tests out.

tobrun commented:
> However we don't usually test assertions 

👍 GTK, removed that commit and cleaned the others.

while writing the tests I noticed that we only have render test for annotations. Do we want to be able to test the business logic? If yes, what is your opinion on adding code to be able unit test them? (we currently don't expose a getters for symbol/shape annotations).

jfirebaugh commented:
Is there functionality we could test that isn't covered by the rendering tests, and could we test it without adding accessors? In general, the rendering tests work well for us because they test fairly stable APIs and don't depend on adding accessors just for the sake of testing.

tobrun commented:
We are currently testing the overall feature through UI tests. While these tests detect a feature being broken, there are scenarios that changes to business logic would remain undetected. 

An example for this in this PR. We changed the internal logic of annotation manager but these changes did not reflect in any core tests failing. The java tests did however [flagged](https://www.bitrise.io/build/3e6606401ed2b475) some changes. 

For me, coming from another language and not being confident about changes introduced, writing unit tests help me validate my assumptions and help me remove the fear of changing code. Those are my 2 cents on the topic. I know this fits into a larger picture discussion about public API design and how you apply testing to it. It also comes close to the mistake of testing framework internals versus testing our own code.

jfirebaugh commented:
> We changed the internal logic of annotation manager but these changes did not reflect in any core tests failing.

They did, actually -- 722053c61264e5998506c8e47d1685071a03bbec was failing in CI before the additional changes to `AnnotationManager`.

> The java tests did however flagged some changes.

Can you say more about what exactly they flagged?

tobrun commented:
> Can you say more about what exactly they flagged?

They flagged misuse in the test code.. they did not flag the business logic change..

For me, testing is validating if the code works and thus also capturing the edge cases in tests. What we have now is that these cases are now encapsulated by doing runtime tests (assertions). 

Without extra 👀  I would have written:
 - addAnnotationTest
 - addRemoveAnnotationTest
 - updateAnnotationTest
 - updateNotExisitingAnnotationTest - expect assert
 - ...

I'm taking notes here and seeing that it can be done differently than what I'm used to. A benefit from current approach is that they test the feature as a black box and these will sustain change better than my white box testing approach.

The only remark is that I feel I'm still required to run the bindings to verify my changes. While with the white box testing approach I would feel confident doing this without.
